pkg/agents/team4/finalAgent/agent.go

```diff
@@ -0,0 +1,189 @@
+	}
+	daysToWait := map[string][]int{
+		"selfish":  {0, 1, 4, 3}, // TODO: to optimise more int(baseAgent.HealthInfo().MaxDayCritical / 2)
+		"selfless": {0, 1, 4, 3}, // TODO: to optimise more
+	}
+
+	return CustomAgentEvoParams{ //initialise the parameters of the agent
+		locked:             true,
+		foodToEat:          foodToEat,
+		daysToWait:         daysToWait,
+		currentPersonality: "selfless",
+		ageLastEaten:       0,
+		morality:           100 * rand.Float64(), // TODO: Use this properly
+		traumaScaleFactor:  1,
+		healthStatus:       3,
+		globalTrust:        0.0,
+		coefficients:       []float64{2, 4, 8}, // TODO: maybe train these co-efficients using evolutionary algorithm
+		lastFoodTaken:      0,
+		sentMessages:       []messages.Message{},
+		responseMessages:   []messages.Message{},
+		lastPlatFood:       -1,
+		maxFoodLimit:       50,
+		messageCounter:     0,
+		globalTrustLimit:   75,
+		lastAge:            0,
+	}
+}
+
+func New(baseAgent *infra.Base) (infra.Agent, error) {
+	//create other parameters
+	return &CustomAgentEvo{
+		Base:   baseAgent,
+		params: InitaliseParams(baseAgent),
+	}, nil
+}
+
+// Checks if neighbour below has eaten
+func (a *CustomAgentEvo) NeighbourFoodEaten() food.FoodType {
+	if a.CurrPlatFood() != -1 {
+		if !a.PlatformOnFloor() && a.CurrPlatFood() != a.params.lastPlatFood {
+			return a.params.lastPlatFood - a.CurrPlatFood()
+		}
+		return 0
+	}
+	return -1
+}
+
+// removes a specific message from a message array
+func remove(slice []messages.Message, s int) []messages.Message {
+	return append(slice[:s], slice[s+1:]...)
+}
+
+// Check if a day has passed
+func (a *CustomAgentEvo) HasDayPassed() bool {
+	if a.Age() != a.params.lastAge {
+		a.params.lastAge = a.Age()
+		return true
+	}
+	return false
+}
+
+func (a *CustomAgentEvo) Run() {
+	a.Log("Reporting agent state of team 4 agent", infra.Fields{"health": a.HP(), "floor": a.Floor()})
+
+	if a.CurrPlatFood() != a.params.lastPlatFood && a.PlatformOnFloor() {
+		a.params.lastPlatFood = a.CurrPlatFood()
+	}
+
+	receivedMsg := a.ReceiveMessage()
+	if receivedMsg != nil {
+		receivedMsg.Visit(a)
+	} else {
+		a.Log("I got nothing")
+	}
+
+	//TODO: Define a threshold limit for other agents to respond to our sent message.
+	a.SendingMessage()
+
+	// a.intendedFoodTaken = food.FoodType(int(int(a.CurrPlatFood()) * (100 - int(a.globalTrust)) / 100))
+	// a.lastFoodTaken, _ = a.TakeFood(a.intendedFoodTaken)
+
+	//dayPass := a.HasDayPassed()
+
+	healthLevelSeparation := int(0.33 * float64(a.HealthInfo().MaxHP-a.HealthInfo().WeakLevel))
+
+	if a.HP() <= a.HealthInfo().WeakLevel { //critical
+		a.params.healthStatus = 0
+		a.params.locked = false
+		// if dayPass {
+		// 	a.params.traumaScaleFactor = math.Min(200, a.params.traumaScaleFactor+0.03)
+		// }
+	} else if !a.params.locked {
+		if a.HP() <= a.HealthInfo().WeakLevel+healthLevelSeparation { //weak
+			a.params.healthStatus = 1
+			a.params.locked = true
```

Comment:
```suggestion
		a.params.locked = true
		if a.HP() <= a.HealthInfo().WeakLevel+healthLevelSeparation { //weak
			a.params.healthStatus = 1
```
Then you don't need to copy that line everywhere.